### PR TITLE
Wrapped library with familiar _.function convention.

### PR DESCRIPTION
**Pull request by kmalakoff**
Hi Robert,

Yesterday, I went looking for an objective-C implementation of underscore and found yours! I have very impressed with the amazing property trick for implementing .functionName syntax. 

However, felt the syntax and forced chaining could a little unconventional in the sense that I believe an Underscore.js user would just want something identical as much as possible. So I wrapped your library using the _.function convention, I introduced a convention of _.functionNameC for the library user to choose Chaining, and I added _,indexOf and _.defer because I was using them in my own library.

A few thoughts:

1) As an optimization, I copied some common code to the _.functionName implementation. Obviously having duplicate code is not great, but I wanted to avoid unnecessary collection copying through the Wrappers. Maybe someone in the community can review this for more optimizations (maybe some implementations belong in the Wrappers and some in the _.functionName).

2) I was originally trying to store _ in an extern rather than as a singleton, but couldn't get it to work. If you have any ideas to do that to avoid the extra [Underscore sharedInstance] calls, please do!

3) I ended up removing some of your '- (id)init __deprecated;' for the singleton to create an instance. Hopefully that is the best approach.

I hope you accept this pull request!

Kevin

### CODE REVIEW

**Comment by kmalakoff** [comment]
I realized that with _.wrapArray and _.wrapDict, all of the _.functionNameC become redundant. Updated library and tests to reflect this.

**Comment by robb** [comment]
Hi Kevin,

thanks _a lot_ for the pull request! I'll have a look :-)

File: Underscore/USCollectionWrapper.m [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+#import "USCollectionWrapper.h"
+
+@implementation USCollectionWrapper
+
+#pragma mark Lifecycle
+
+- (id)unwrap;
+{
+  NSAssert(nil, @"USCollectionWrapper unwrap is abstract and should not be called");
```
Comment (robb) [inline Underscore/USCollectionWrapper.m]: If `-[USCollectionWrapper unwrap]` is not meant to be called anyway, what is the reasoning behind having a common superclass?

`-[USArrayWrapper unwrap]` and `-[USDictionaryWrapper unwrap]` have different return types (NSArray\* and NSDictionary*, respectively)

File: Underscore/Underscore+Collection.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+//  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+//  IN THE SOFTWARE.
+//
+
+#import "Underscore.h"
+
+@interface Underscore (Collection)
+
+@property (readonly) id(^each)(id collection, id block);
+@property (readonly) id(^map)(id collection, id block);
```
Comment (robb) [inline Underscore/Underscore+Collection.h]: I don't like that this takes away the static typing,

```
_.each(@[@"foo", @"bar", @"baz"], ^(id key, id obj) {
    // ...
});
```

will - unlike JavaScript - give you a nasty runtime crash. I think we should use static types throughout Underscore.m.
Have you considered e.g. `_.arrayEach` ?

File: Underscore/USCollectionWrapper.m (hunk 1 above)
Comment (kmalakoff) [inline Underscore/USCollectionWrapper.m]: I'm not sure the best way to do this. The problem is that there are two _.each() functions...aliasing works fine for _.each() versions, but the wrapped versions are a little problematic since the next step in the chain will require a cast from USCollectionWrappers to the actual type to continue the chain...

What I would recommend in this case is to make _.each return void (like in underscore) and drop the NSArrayWrapper and NSDictionaryWrapper implementations since the returning a wrapped sort of doesn't make sense, and then get rid of NSCollectionWrapper and the NSArrayWrapper and NSDictionaryWrapper implementations.

Basically, I put it in there to try to maintain some sort of backwards compliance to the Wrappers, but it is a bit of a stretch....

File: Underscore/Underscore+Collection.h (hunk 2 above)
Comment (kmalakoff) [inline Underscore/Underscore+Collection.h]: That is an option, but it isn't compliant with Underscore.js. See my other comment about dropping the Wrapper versions. 

I think that in this case, the compliance is worth the runtime risk. FYI: I was thinking about adding an assert if the collection and block types didn't match. Maybe that's a good safeguard?

**Comment by kmalakoff** [comment]
OK. Here's my proposal to deal with each and UICollectionWrapper...

File: Underscore/Underscore+Collection.h (hunk 2 above)
Comment (robb) [inline Underscore/Underscore+Collection.h]: Unfortunately, I don't think there is a way to find out a blocks arity at runtime

**Comment by kmalakoff** [comment]
Or better yet...just returning void.

BTW: it seemed impossible to check the block signature at runtime: http://stackoverflow.com/questions/9048305/checking-objective-c-block-type

File: Underscore/Underscore+Collection.h (hunk 2 above)
Comment (kmalakoff) [inline Underscore/Underscore+Collection.h]: Yes, that appears to be the case :-(  although we can check the collection type...

(http://stackoverflow.com/questions/9048305/checking-objective-c-block-type)

So void return type (see latest pull request update) and assuming that people exhaustively test their code at runtime to catch mis-matches is the best tradeoff? or you like _.eachArray and _.eachDict? (I prefer compliance with underscore)

Comment (robb) [inline Underscore/Underscore+Collection.h]: I think I prefer type safety over compliance with JavaScript :-)

I've started work to work locally now, I'll show you my results in a bit

**Comment by robb** [comment]
Have a look at https://github.com/robb/Underscore.m/compare/24206ef6a1a6d6b7f4de48310679b041d352f024...shortcuts

I started messing with the preprocessor as I was trying to get things a bit more DRY. I failed :smile:

Then I rewrote the old test to use the new `_.function(foo)` syntax as this uses wrapping internally anyways and I don't see why we should test the same code twice*.

Then I also went for `_.arrayEach` and `_.dictEach` respectively, both returning `void`.
The old `_.array(foo).each` still works as before for chaining.

I left out `indexOf` and `defer` as I think they should be introduced separately. I like `indexOf` and I have to think a bit about `defer`.

Instead of having a `sharedInstance`, I made the "shortcuts" class methods. Now people can define _ to be Underscore and also get `_.isArray` without extra block calls.

Is anything missing that I didn't address?
- famous last words

File: Underscore/Underscore+Collection.h (hunk 2 above)
Comment (kmalakoff) [inline Underscore/Underscore+Collection.h]: Hmmm...I guess I totally disagree ;-)  

If I'm going to use a library called Underscore, it should try to have the same signature as the original if possible: 

```
_.each([1, 2, 3], function(num){ alert(num); });
=> alerts each number in turn...
_.each({one : 1, two : 2, three : 3}, function(num, key){ alert(num); });
=> alerts each number in turn...
```

If you really want to supply a safe version, I would provide both _.eachArray/_.eachDict and _.each without block type check. If people are unhappy with the lack of type safety on _.each, they can use the safe version. 

Give them the choice! (personally, I would not use _.eachArray and _.eachDict).

That said, people can always branch and extend their own versions with categories.

**Comment by kmalakoff** [comment]
Interesting. I guess making them class methods works too! (and it gets rid of the singleton!) 

1) I recommend you should put '#define _ Underscore' in the header to give signature compliance with Underscore.js.

2) Because people can extend Underscore using categories, I've been thinking about providing an extension mechanism...The problem I currently have is that I've added my own categories on the end of Underscore.h. What do you think about creating a dummy file called UnderscoreExtensions.h where people can include their own headers for their categories, but that means the file will be missing from the library on first use. Any better ideas? (my thought is that if they upgrade their version of the library, they shouldn't have to update their Underscore.h file -> alternatively, people can just create their own header where they import Underscore.h, but just wondering if there is a way to allow people to import Underscore.h in their .m files AND add extensions)

3) Personally, I like categories that are more descriptive like in the original implementation since they are not "Shortcuts" but some will be the native implementation. How about following Underscore.js (http://underscorejs.org/) left column categories: Underscore+Collections, Underscore+Arrays, Underscore+Functions, Underscore+Objects, Underscore+Utility, and Underscore+Chaining. This will give the community something to aim at in filling in Underscore.m! Here are some Underscore+Functions to get you started:

```
@interface Underscore (Functions)

+ (void(^)(UnderscoreDoBlock block))defer;
+ (void(^)(UnderscoreDoBlock block, NSUInteger wait))delay;

@end
```

```
@implementation Underscore (Functions)

+ (void(^)(UnderscoreDoBlock))defer 
{ 
  return ^(UnderscoreDoBlock block) {
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0), dispatch_get_main_queue(), block);
  };
}

+ (void(^)(UnderscoreDoBlock, NSUInteger))delay
{
  return ^(UnderscoreDoBlock block, NSUInteger wait) {
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, wait), dispatch_get_main_queue(), block);
  };
}

@end
```

4) It looks like you got rid of your tests for Wrappers! The wrappers are great...don't drop them! I would put them back and go back to having separate files for tests and categories like in my original pull request. Also, looking at underscore's chaining:

```
chain_.chain(obj) 
Returns a wrapped object. Calling methods on this object will continue to return wrapped objects until value is used.

var stooges = [{name : 'curly', age : 25}, {name : 'moe', age : 21}, {name : 'larry', age : 23}];
var youngest = _.chain(stooges)
  .sortBy(function(stooge){ return stooge.age; })
  .map(function(stooge){ return stooge.name + ' is ' + stooge.age; })
  .first()
  .value();
=> "moe is 21"
```

maybe you want to put your USArrayWrapper and USDictionaryWrapper in Underscore+Chaining and implement _.chainArray and _.chainDictionary (and maybe a non-typesafe version like _.chain where the caller needs to cast)? (and maybe call your classes USArrayChainer and USDictionaryChainer).

In short, I like your idea of using class methods, but in other areas, I think there are still some great improvements that should be made!

Cheers,

Kevin

PS. Since I spent quite a bit of time to create the initial version (I think you only needed to replace - with +) and to submit ideas which I think will make Underscore.m much better, can you please give me an honorable mention (It can be as small as "Special thank you to Kevin Malakoff (https://github.com/kmalakoff)") with a link to my GitHub profile somewhere? (It doesn't have to be big)

**Comment by kmalakoff** [comment]
OK. I've just implemented all of the points above except for the extension mechanism. 

This is a major change...let me know what you think!

**Comment by stigi** [comment]
hi @kmalakoff,
quite a pull request you created there (2,723 additions and 426 deletions) :)

I think this PR is mixing up to many different concerns that it will ever be merged in total.
Do you think you could split it up in multiple pull requests so it's possible to discuss the different aspecs of it?
Especially the renaming of `*Wrapper` to `*Chainer` and changing things in the API is something that you should not assume to be merged in without being previously discussing it. Having a pull request with the suggested changes attached to this discussion might though be helpful.

**Comment by kmalakoff** [comment]
@stigi : Hmmm. I really don't think multiple pull requests is a good way to analyze the changes since they are too far reaching. I've already spent a good portion of the weekend on this so first read my explanation and if you see believe this is needed, please let me know.

Fortunately, what I am proposing is very simple. Make Underscore.m more like Underscore.js!

That means:

1) Reorganize the library around the same categories as Underscore.js. I've used ObjectiveC categories to match the ones in Underscore.js.  

-> Look at the left column categories on http://underscorejs.org/ and they match the categories in the refactored version Underscore+Collections, Underscore+Arrays, Underscore+Functions, Underscore+Objects, Underscore+Utility, and Underscore+Chaining

2) Underscore.m was originally written to implement _.chain() in Underscore.js, but I think the author didn't realize it or focussed too much on chaining and not enough on the full Underscore library. So basically, I depreciated the _array and _dict (replacing them with _.chainArray() and _.chainDict() to better match Underscore.js) and hid away all of the implementation details inside USArrayChainer and USDictionaryChainer (renamed because Wrapper doesn't say why it is wrapped, but Chainer does give a nice noun to be more self-evident!).

-> besides putting the Chaining stuff in a folder, I also moved some of the chaining functionality from NSDictionaryWrapper to Underscore+Objects to make them better correspond to Underscore.js categories 

3) Quickly moved some things under _.functionName to be more compliant with Underscore.js's functions and to allow users to start with _.functionName rather than the chaining syntax which I believe matches a large portion of the use cases (I haven't yet found a use for chaining and just need a simple, optimized, unchained version). Also, this well-positions the library with ObjectiveC categories for future expansion. There's the whole left sidebar on http://underscorejs.org/ to fill in (I submitted _delay and _.defer to get you started).

-> What I think you'll want to do is to move some of the functionality from the Chainers to the categories (currently, I just did a quick bridge from _.functionName to the Chainers while reimplementing some simple ones in the _.functionName so the job still requires a bit more work to flip it so that the Chainers reference the _.functionName implementations rather than the other way round).

Makes sense? Simple?!

Kevin

**Comment by robb** [comment]
I think I am with @stigi here, let's not get ahead of ourselves.

There are multiple things being proposed here:
- Introducing a `Underscore.method` syntax as an alternative to wrapping and unwrapping  
  (This poses the question how to handle `Underscore.each` in the context of static typing)
- Removing the old `_array` and `_dict` macros in favor of the new syntax' `Underscore.array` and `Underscore.dict`
- Aliasing `_` to `Underscore`
- Adding an `indexOf` method
- Adding an `defer` method
- Restructuring Underscore.m into a couple of categories  
  (Is this a viable way for app-specific extensions, too?)
- Closely aligning Underscore.m with underscore.js
- Renaming the Wrappers into "Chainers"

I strongly believe that all those points need to be discussed separately. I'm not sure I'm in line with all of the changes introduced by this PR and at this point and it's hard to cherry-pick individual aspects.

Let's start with the first item in the list, the `Underscore.method` syntax:
I'm generally in favor of something along the lines of https://github.com/robb/Underscore.m/compare/819b79f7245a1a7e75f544cf9d1303930df1e16c...4d4b8dd400b3214a340a160720dbe2c065c1c747
but I'd like to discuss its implications on e.g. a future support of `NSSet`…

**Comment by kmalakoff** [comment]
Understood. I wish it was easy to cherry pick, but it is not really. I just think you should start with the fundamental question "Make Underscore.m more like Underscore.js" or not. From that, all these points naturally get resolved.

Quick feedback to clarify my proposals:

```
1. Introducing a Underscore.method syntax as an alternative to wrapping and unwrapping
```

-> I would clarify "Introducing a Underscore.method syntax and migrating code from wrapping to Underscore". In other words, chaining would reuse Underscore.m like like in Underscore.js (not the other way around as it is now):

```
_.chain = function(obj) {
    return _(obj).chain();
  };
```

```
2. Removing the old _array and _dict macros in favor of the new syntax' Underscore.array and Underscore.dict
```

-> Actually, the proposal is Underscore.chainArray and Underscore,chainDict to be compliant with Underscore.js as in 1 (_.chain).

```
4. Adding an indexOf method
5. Adding an defer method
```

-> I would wrap these together (plus add _.delay) and call it "Fully implementing Underscore.js in Underscore.m (_.indexOf, _.defer, _.delay, etc)". 

-> I also included in Helpers like your negate, a couple of others that are useful in objectiveC (propTester and propStringTester).

```
6. Restructuring Underscore.m into a couple of categories
```

-> The proposal is "Structuring Underscore.m into all of the Underscore.js categories: Collections, Arrays, Functions, Objects, Utility, and Chaining and adding the Helpers category for ObjectiveC-specific extensions"

```
8. Renaming the Wrappers into "Chainers"
```

-> I would say it is "Making Underscore.m consistent with Underscore.js on the terms 'wrap' and 'chain'. For example:

Wrap:

```
wrap_.wrap(function, wrapper) 
Wraps the first function inside of the wrapper function, passing it as the first argument. This allows the wrapper to execute code before and after the function runs, adjust the arguments, and execute it conditionally.

var hello = function(name) { return "hello: " + name; };
hello = _.wrap(hello, function(func) {
  return "before, " + func("moe") + ", after";
});
hello();
=> 'before, hello: moe, after'
```

Chain:

```
chain_.chain(obj) 
Returns a wrapped object. Calling methods on this object will continue to return wrapped objects until value is used.

var stooges = [{name : 'curly', age : 25}, {name : 'moe', age : 21}, {name : 'larry', age : 23}];
var youngest = _.chain(stooges)
  .sortBy(function(stooge){ return stooge.age; })
  .map(function(stooge){ return stooge.name + ' is ' + stooge.age; })
  .first()
  .value();
=> "moe is 21"
value_(obj).value() 
Extracts the value of a wrapped object.

_([1, 2, 3]).value();
=> [1, 2, 3]
```

So I would go a step further and suggest 'unwrap' be changed to 'value' as well for consistency sake (I've updated my pull request). On further thought, I guess it works either way Wrapper or Chainers (or there might be a better term!). I've reverted to Wrapper in my branch since Underscore.js does use the word wrapper for functions and the wrapped version of underscore that is returned...I wonder if there is a way to do the same thing in Underscore.m...maybe:

```
@interface UnderscoreWrapper : Underscore

@property  id *value;

@end

@interface Underscore (Chaining)

+ (UnderscoreWrapper *(^)(id obj))chain;

@end

```

and then do all of the chaining using runtime checks on the obj using 'isKindOfClass'. Possible? Interesting?

Good luck with your debates! If you want to involve me, just let me know.

Kevin

**Comment by kmalakoff** [comment]
OK guys. You are not going to be happy (or maybe very happy!), but I moved all the chaining to be general purpose and runtime checks and made it compliant with Underscore.

Here is Underscore.js:

```
var stooges = [{name : 'curly', age : 25}, {name : 'moe', age : 21}, {name : 'larry', age : 23}];
var youngest = _.chain(stooges)
  .sortBy(function(stooge){ return stooge.age; })
  .map(function(stooge){ return stooge.name + ' is ' + stooge.age; })
  .first()
  .value();
=> "moe is 21"
```

and here is Underscore.m:

```
  NSArray *stooges = [NSArray arrayWithObjects:
    [NSDictionary dictionaryWithObjectsAndKeys: @"curly", @"name", [NSNumber numberWithInt:25], @"age", nil], 
    [NSDictionary dictionaryWithObjectsAndKeys: @"moe", @"name", [NSNumber numberWithInt:21], @"age", nil], 
    [NSDictionary dictionaryWithObjectsAndKeys: @"larry", @"name", [NSNumber numberWithInt:23], @"age", nil], 
    nil];

  NSString *youngest = _.chain(stooges)
    .sortBy(^(NSDictionary *stooge){ return [stooge valueForKey:@"age"]; })
    .map(^(NSDictionary *stooge, id key, id collection){ 
      return [NSString stringWithFormat:@"%@ is %@", [stooge valueForKey:@"name"], [stooge valueForKey:@"age"]]; 
    })
    .first()
    .value();

  STAssertTrue([youngest isEqualToString:@"moe is 21"], @"stooges comply with Underscorejs");
```

@robb : this also means I made a proper, sharable iterator (that is compliant with Underscore.js minus the optional arguments that JavaScript brings) so we don't need _.eachArray, etc and everything is type safe. The new definition is:

```
typedef void (^USUnderscoreIteratorBlock)(id value, id key, id collection);
typedef id   (^USUnderscoreMapBlock)(id value, id key, id collection);
```

Maybe rather than trying to pick through the many changes, you want to consider taking everything whole-hearted and moving Underscore.m in the direction of being as similar as possible to Underscore.js? Just a thought...

Cheers,

Kevin

**Comment by kmalakoff** [comment]
I updated the README to give a succinct overview of where I'm hoping Underscore.m will move to: https://github.com/kmalakoff/Underscore.m/blob/master/README.md

Enjoy! (I'm quite excited by the future possibilities)

**Comment by robb** [comment]
Hi Kevin,

to make a long story short, I don't think that Underscore.m needs to follow underscore.js precisely.
Objective-C and Javascript are vastly different languages and we need to carefully evaluate what parts of underscore.js make sense on iOS and OS X.

Also, this pull request far surpassed a number of changes that I would be comfortable with merging.
According to a quick `sloccount`, It more than doubles the code size compared to 819b79f7245a1a7e75f544cf9d1303930df1e16c.

Don't get me wrong, you made some very valid points but I think they need separate discussions.
I'm opening a couple of Issues, feel free to raise some more :-)

**Comment by kmalakoff** [comment]
Yes, it is basically a total re-write of the Underscore.m library. I guess I just think that focussing on the chaining mechanics exclusively isn't the best approach.

Luckily with GitHub, people will have a choice to use yours or mine!

Good luck in the future!

Kevin

**Comment by robb** [comment]
> I guess I just think that focussing on the chaining mechanics exclusively isn't the best approach.

I agree, see #16

**Comment by kmalakoff** [comment]
I'm sorry, but I'm just not very interested in dissecting the changes especially if they are already in the standard (for example, Lo-Dash is Underscore.js compliant so it can be exchanged)...what I'd like to focus on is making Underscore.m as compliant as possible with Underscore.js! (despite the differences in language)

What I would prefer is if you take my changes fully so we have a common platform ;-)

A glimpse of what I'm thinking about: 1) make a table of the Underscore.js and Underscore.m functions 2) have a column to note the API differences where they exist 3) start filling the table in and eliminating incompatibilities 4) port the tests from Underscore.js into ObjectiveC. 

If you share this future for Underscore in ObjectiveC...wanna join forces?

**Comment by robb** [comment]
I don't think it's necessary to maintain API compatibility with a library written in JavaScript that is commonly run in the browser or node.js. Why throw the static typing in Objective-C out of the window just because there's none in JavaScript?

Underscore.m is meant to be a useful utility library for Objective-C developers, being 100% compatible with underscore.js does not help in that regard.

However, there are still quite a few functions in underscore.js that should be included in Underscore.m, such as `indexOf` and `zip`.
You raised some very valid points and I'm more than happy to discuss them in their respective issues/pull requests
(e.g. #15 and #16 – feel free to add some more).

**Comment by kmalakoff** [comment]
@robb : I totally understand where you are coming from. I have a statically type safe iterator and map blocks, but all my chaining is dynamically type checked. I'm happy with the tradeoff since it allows for a more complete, compatible library. 

To be clear, I'm not advocating for 100% compatibility because that is impossible ;-) What I'd like to do is create an environment where 1) people can copy and paste their JavaScript code, make a few modifications, and run it in Objective-C, 2) people can use Underscore.js documentation and a compatibility table to learn the Objective-C version.

As I mentioned earlier in the pull request, first we need to agree on the goal...

   Yours: Small, lightweight underscore chaining library for Objective-C.

   Mine: Complete, extensible (through categories), as-compatible-as-possible underscore implementation for Objective-C.

So it sounds like unfortunately, we won't be able to get in sync on this and since our goals are different, I think the best thing is for me to finish my re-write and to release my library under a different name to avoid confusion. Agreed?

**Comment by kmalakoff** [comment]
Here's the first release: https://github.com/kmalakoff/_.m
